File: frontend/lib/core/models/observation.dart
import 'site.dart';
import 'user.dart';

class Observation {
  final int? id;
  final DateTime? observationDate;
  final int? siteId;
  final Site? site;
  final int? createdById;
  final User? createdBy;
  final double? temperature;
  final double? humidity;
  final int? workersPresent;
  final int? hoursWorked;
  final List<String>? activities;
  final double? epiCompliance;
  final double? equipmentState;
  final double? siteClean;
  final double? fatigueLevel;
  final int? minorIncidents;
  final int? majorIncidents;
  final String? notes;
  final List<String>? photoUrls;
  final String? aiAnalysis;
  final double? riskScore;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Observation({
    this.id,
    this.observationDate,
    this.siteId,
    this.site,
    this.createdById,
    this.createdBy,
    this.temperature,
    this.humidity,
    this.workersPresent,
    this.hoursWorked,
    this.activities,
    this.epiCompliance,
    this.equipmentState,
    this.siteClean,
    this.fatigueLevel,
    this.minorIncidents,
    this.majorIncidents,
    this.notes,
    this.photoUrls,
    this.aiAnalysis,
    this.riskScore,
    this.createdAt,
    this.updatedAt,
  });

  factory Observation.fromJson(Map<String, dynamic> json) {
    return Observation(
      id: json['id'],
      observationDate: json['observationDate'] != null
          ? DateTime.parse(json['observationDate'])
          : null,
      siteId: json['site'] != null ? json['site']['id'] : json['siteId'],
      site: json['site'] != null ? Site.fromJson(json['site']) : null,
      createdById:
          json['createdBy'] != null ? json['createdBy']['id'] : json['createdById'],
      createdBy:
          json['createdBy'] != null ? User.fromJson(json['createdBy']) : null,
      temperature: json['temperature'] != null
          ? (json['temperature'] is double
              ? json['temperature']
              : double.tryParse(json['temperature'].toString()))
          : null,
      humidity: json['humidity'] != null
          ? (json['humidity'] is double
              ? json['humidity']
              : double.tryParse(json['humidity'].toString()))
          : null,
      workersPresent: json['workersPresent'],
      hoursWorked: json['hoursWorked'],
      activities: json['activities'] != null
          ? List<String>.from(json['activities'])
          : null,
      epiCompliance: json['epiCompliance'] != null
          ? (json['epiCompliance'] is double
              ? json['epiCompliance']
              : double.tryParse(json['epiCompliance'].toString()))
          : null,
      equipmentState: json['equipmentState'] != null
          ? (json['equipmentState'] is double
              ? json['equipmentState']
              : double.tryParse(json['equipmentState'].toString()))
          : null,
      siteClean: json['siteClean'] != null
          ? (json['siteClean'] is double
              ? json['siteClean']
              : double.tryParse(json['siteClean'].toString()))
          : null,
      fatigueLevel: json['fatigueLevel'] != null
          ? (json['fatigueLevel'] is double
              ? json['fatigueLevel']
              : double.tryParse(json['fatigueLevel'].toString()))
          : null,
      minorIncidents: json['minorIncidents'],
      majorIncidents: json['majorIncidents'],
      notes: json['notes'],
      photoUrls: json['photoUrls'] != null
          ? List<String>.from(json['photoUrls'])
          : null,
      aiAnalysis: json['aiAnalysis'],
      riskScore: json['riskScore'] != null
          ? (json['riskScore'] is double
              ? json['riskScore']
              : double.tryParse(json['riskScore'].toString()))
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'observationDate': observationDate?.toIso8601String().split('T')[0],
      'site': siteId != null ? {'id': siteId} : site?.toJson(),
      'temperature': temperature,
      'humidity': humidity,
      'workersPresent': workersPresent,
      'hoursWorked': hoursWorked,
      'activities': activities,
      'epiCompliance': epiCompliance,
      'equipmentState': equipmentState,
      'siteClean': siteClean,
      'fatigueLevel': fatigueLevel,
      'minorIncidents': minorIncidents,
      'majorIncidents': majorIncidents,
      'notes': notes,
      'photoUrls': photoUrls,
      'aiAnalysis': aiAnalysis,
      'riskScore': riskScore,
    };
  }
}

File: backend/src/main/java/com/safesite/model/Observation.java
package com.safesite.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "observations")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Observation {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(name = "observation_date", nullable = false)
    private LocalDate observationDate;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "site_id", nullable = false)
    private Site site;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by", nullable = false)
    private User createdBy;
    
    // Weather & Presence
    private Double temperature;
    private Double humidity;
    private Integer workersPresent;
    private Integer hoursWorked;
    
    // Activities
    @ElementCollection
    @CollectionTable(name = "observation_activities", joinColumns = @JoinColumn(name = "observation_id"))
    @Column(name = "activity")
    private List<String> activities = new ArrayList<>();
    
    // Safety Evaluations
    @Column(name = "epi_compliance")
    private Double epiCompliance; // 0-100
    
    @Column(name = "equipment_state")
    private Double equipmentState; // 0-10
    
    @Column(name = "site_clean")
    private Double siteClean; // 0-10
    
    @Column(name = "fatigue_level")
    private Double fatigueLevel; // 0-10
    
    // Incidents
    @Column(name = "minor_incidents")
    private Integer minorIncidents;
    
    @Column(name = "major_incidents")
    private Integer majorIncidents;
    
    // Notes & Photos
    @Column(columnDefinition = "TEXT")
    private String notes;
    
    @ElementCollection
    @CollectionTable(name = "observation_photos", joinColumns = @JoinColumn(name = "observation_id"))
    @Column(name = "photo_url")
    private List<String> photoUrls = new ArrayList<>();
    
    // AI Analysis Result
    @Column(name = "ai_analysis", columnDefinition = "TEXT")
    private String aiAnalysis;
    
    @Column(name = "risk_score")
    private Double riskScore;
    
    @Column(name = "created_at")
    private LocalDateTime createdAt;
    
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
    
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (observationDate == null) {
            observationDate = LocalDate.now();
        }
    }
    
    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

